core/example/solvers/pme.cpp

```diff
@@ -0,0 +1,449 @@
+#include "definitions.h"
+#include "pme.h"
+#include <cmath>
+#include <fftw3.h>
+#include <time.h>
+#include <sys/time.h>
+
+/* Smooth particle mesh Ewald (SPME) solver
+ * - This method, from Essman et al. (1995) computes long-range Coulombic forces
+ *   with O(nlogN) scaling by using 3D FFT and interpolation to a mesh for the 
+ *   reciprocal space part of the Ewald sum.
+ * - Here the method is used to compute electrostatic energies from an example
+ *   arrangement of charged particles. Currently, we assume periodic boundary conditions
+ *   and a cubic mesh and arrangement of particlesi in 3 dimensions.
+ * - Future versions will allow for assymetric meshes and non-uniform particle
+ *   distributions, as well as 1 or 2 dimensions. 
+ */
+
+
+//constructor given an accuracy
+TPME::TPME(double accuracy, ParticleList particles, double lx, double ly, double lz)
+{
+  _r_max = 0.0;
+  tune(accuracy,particles,lx, ly, lz);
+}
+
+//set base values for alpha, r_max, k_max
+TPME::TPME(double alpha, double r_max, double k_max)
+{
+  _alpha = alpha;
+  _r_max = r_max;
+  _k_max = k_max;
+  _k_max_int[0] = std::ceil(_k_max);
+  _k_max_int[1] = std::ceil(_k_max);
+  _k_max_int[2] = std::ceil(_k_max);
+}
+
+//Tune to a given accuracy
+void TPME::tune(double accuracy, ParticleList particles, double lx, double ly, double lz)
+{
+  typedef Kokkos::MinLoc<double, int> reducer_type;
+  typedef reducer_type::value_type value_type;
+  value_type error_estimate;
+
+  auto q = particles.slice<Charge>();
+  double q_sum;  
+
+  const int N_alpha = 200;
+  const int N_k = 2000;
+
+  const int n_max = particles.size();
+
+  // calculate sum of charge squares
+  Kokkos::parallel_reduce( n_max, KOKKOS_LAMBDA( const int idx, double& q_part )
+      {
+      q_part += q(idx) * q(idx);
+      },
+      q_sum
+      );
+
+  double r_max = _r_max = std::min(0.49*lx,0.1*lx + 1.0);//real space cutoff value
+  Kokkos::parallel_reduce( "MinLocReduce", N_alpha*N_k,
+    KOKKOS_LAMBDA (const int& idx, value_type& team_errorest)
+    {
+      int ia = idx%N_alpha;
+      int ik = idx/N_alpha;
+
+      double alpha = (double)ia * 0.05 + 1.0;//splitting parameter
+      double k_max = (double)ik * 0.05;//max k-vector
+      
+      //Compute real part of error estimate
+      double delta_Ur = q_sum * 
+      sqrt(0.5 * r_max / (lx * ly * lz )) *
+      std::pow(alpha*r_max,-2.0) *
+      exp(- alpha * alpha * r_max * r_max);
+
+      //Compute k-space part of error estimate
+      double delta_Uk = q_sum * alpha / PI_SQ *
+      std::pow(k_max,-1.5) *
+      exp(-std::pow(PI * k_max / (alpha * lx),2));
+
+      //Total error estimate
+      double delta = delta_Ur + delta_Uk;
+      Kokkos::pair<double,double> values(alpha,k_max);
+
+      if ( (delta < team_errorest.val) && (delta > 0.8 * accuracy))
+      {
+      //  std::cout << alpha << " " << k_max << " " << delta_Ur << " " << delta_Uk << " " << delta << " " << accuracy << std::endl;      
```

Comment:
There are a number of commented out print statements in the code - maybe hide them in a debug diagnostic macro instead? That way you can just set the diagnostic variable once and get output.